CUDADataFormats/Common/interface/PortableHostCollection.h

```diff
@@ -0,0 +1,85 @@
+#ifndef CUDADataFormats_Common_interface_PortableHostCollection_h
+#define CUDADataFormats_Common_interface_PortableHostCollection_h
+
+#include <cassert>
+#include <cstdlib>
+
+#include "HeterogeneousCore/CUDAUtilities/interface/host_unique_ptr.h"
+
+namespace cms::cuda {
+
+  // generic SoA-based product in host memory
+  template <typename T>
+  class PortableHostCollection {
+  public:
+    using Layout = T;
+    using View = typename Layout::View;
+    using ConstView = typename Layout::ConstView;
+    using Buffer = cms::cuda::host::unique_ptr<std::byte[]>;
+
+    PortableHostCollection() = default;
+
+    PortableHostCollection(int32_t elements)
+        // allocate pageable host memory
+        : buffer_{cms::cuda::make_host_unique<std::byte[]>(Layout::computeDataSize(elements))},
+          layout_{buffer_.get(), elements},
+          view_{layout_} {
+      // make_host_unique for pageable host memory uses a default alignment of 128 bytes
+      assert(reinterpret_cast<uintptr_t>(buffer_.get()) % Layout::alignment == 0);
+    }
+
+    PortableHostCollection(int32_t elements, cudaStream_t stream)
+        // allocate pinned host memory, accessible by the current device
+        : buffer_{cms::cuda::make_host_unique<std::byte[]>(Layout::computeDataSize(elements), stream)},
+          layout_{buffer_.get(), elements},
+          view_{layout_} {
+      // CUDA pinned host memory uses a default alignment of at least 128 bytes
+      assert(reinterpret_cast<uintptr_t>(buffer_.get()) % Layout::alignment == 0);
+    }
+
+    // non-copyable
+    PortableHostCollection(PortableHostCollection const&) = delete;
+    PortableHostCollection& operator=(PortableHostCollection const&) = delete;
+
+    // movable
+    PortableHostCollection(PortableHostCollection&&) = default;
+    PortableHostCollection& operator=(PortableHostCollection&&) = default;
+
+    // default destructor
+    ~PortableHostCollection() = default;
+
+    // access the View
+    View& view() { return view_; }
+    ConstView const& view() const { return view_; }
+    ConstView const& const_view() const { return view_; }
+
+    View& operator*() { return view_; }
+    ConstView const& operator*() const { return view_; }
+
+    View* operator->() { return &view_; }
+    ConstView const* operator->() const { return &view_; }
+
+    // access the Buffer
+    Buffer& buffer() { return buffer_; }
+    Buffer const& buffer() const { return buffer_; }
+    Buffer const& const_buffer() const { return buffer_; }
+
+    size_t bufferSize() const { return layout_.metadata().byteSize(); }
+
+    // part of the ROOT read streamer
+    static void ROOTReadStreamer(PortableHostCollection* newObj, Layout const& layout) {
```

Comment:
Well, the CPU-only buffer does not in fact depend on CUDA: building the library does, but writing/reading the collection does not.

If we were keeping the CUDA code around, we could make it so that we conditionally compiles the host-only part also without CUDA.

Reply:
Right, the CPU-only buffer does not use any CUDA functionality, but technically there is a dependence chain from `cms::cuda::PortableHostCollection<T>` to CUDA.

The direct dependence could be avoided by making the deleter polymorphic. That would leave the only dependence on CUDA to come via the library (well, also `make_host_unique()` overload set has similar dependence on CUDA, but that could be avoided by some rearrangement of the code).

What would be the intended use case? Just some testing (like reading in the CUDA-variant-written to an Alpaka-variant SoA)? Or something like writing and reading e.g. tracks in any production-like setting (including IB tests)?

Reply:
> > The direct dependence could be avoided by making the deleter polymorphic. That would leave the only dependence on CUDA to come via the library (well, also `make_host_unique()` overload set has similar dependence on CUDA, but that could be avoided by some rearrangement of the code).
> 
> I was thinking something more simple, like `#ifdef`ing away everything that depends on CUDA if it is not available, leaving only the non-pinned host buffer functionality.

Would this include `#ifdef`ing away all CUDA-dependent code in `CUDADataFormats/Common` and in all packages that define a data product using `PortableHostCollection<T>`?

I think a "proper way" would be to define the `cms::cuda::PortableHostCollection` e.g. in `DataFormats/Portable`, make the `unique_ptr` deleter polymorphic, and move the CUDA-independent `make_host_unique` functions to `DataFormats/Portable` as well (although it's not really ideal placement either). The instantiations of `cms::cuda::PortableHostCollection` would go in `DataFormats`, while the instantiations of `cms::cuda::PortableDeviceCollection` would go in `CUDADataFormats`.

I'm trying to think whether the use cases are something that would "require" the "proper way", or simple and low-risk enough to allow shortcuts (as an exception for the persistence requirements).

Reply:
I assume the goal is still to move away from the CUDA code base within the next 6-12 months - then I would not spend any time making it "future proof".

On the other hand, I still do not see the issue with persistent collections that depend (at build time) on CUDA: if we ever need to read them from a CUDA-less architecture, we think that we know how to make changes to the code base to do it.
Trying to make them pre-emptively, when we already have a better approach, seems a bit of a waste of personpower.

Still, if you think we _should_ do it, it shouldn't take much effort to port the Alpaka approach for that as well - either making the compilation dependent on the CUDA availability, or splitting the packages, as you suggest.